Initial support for CHIME F-Engine packet capture with the upgraded hardware/software.

andrerenard opened the pull request:
- Updates the packet capture code for CHIME to support more than 4 ports. 
- Allows for the use of `chordMetadata`
- Adds a new GPU command, `cudaCopyNtoRingbuffer` to merge a list of buffers with one frequency each into a single ring with 16 frequencies (will require a transpose and reorder kernel to use with the CHIME kernels).
- Cleans up the buffer status logging to allow for per-buffer log levels
- Updates the CPU affinity for the frame zeroing threads to use the config file
- Removes support for versions of DPDK older than 19.11 
- Adds PCIe block-list support to get around issues with Nvidia/Mellanox cards getting automatically included in the list of DPDK enabled cards. 

Known issues/todo:
- When using chordMetadata the zeroing/reset logic happens in the critical packet loop, which is resulting in considerable packet loss.   Hopefully the refactored metadata will improve this.  Although likely this will require reworking the critical packet loop to remove buffer/metadata handling entirely. 
- The packet loss array uses the old CHIME format, and has not been updated to the new PL CHORD+CHIME array format.
- This needs to be integrated with the transpose and reorder kernel before we can start using the new CHIME kernels.

andrerenard committed:
DPDK shuffle updated to handle more than one group oflinks, old DPDK support removed, suport chordMetadata in DPDK, and initial CHIME upgrade test config

andrerenard committed:
Merge branch 'chord' of https://github.com/kotekan/kotekan into ar/chime_network

andrerenard committed:
Merge branch 'chord' of https://github.com/kotekan/kotekan into ar/chime_network

andrerenard committed:
Merge branch 'chord' of https://github.com/kotekan/kotekan into ar/chime_network

jbmertens committed:
change linking

andrerenard committed:
Allow logging in the buffer to be config controlled, and move CPU affinity to config for zeroing threads

andrerenard committed:
Added logic for copying the CHIME voltage data to a GPU ring buffer with chordMetadata

andrerenard committed:
Merge branch 'chord' of https://github.com/kotekan/kotekan into ar/chime_network

andrerenard committed:
clang-format

rhaas80 commented on config/chime_upgrade_network.j2:
```diff
@@ -0,0 +1,208 @@
+  alignment: samples_per_frame * 4
+  # List of PCIe devices we don't want to attach too
+  # Required to exclude the ConnectX 6 Lx output NICs. 
+  pcie_block_list:
```
this list the device by its PICe location, does it? not by some sort of identifier of the device itself (vendor id, serial number)?

andrerenard replied:
Yep it uses PCIe addresses.  Perhaps not ideal, but given our largely uniform hardware, it didn't seem worth doing address lookups on some other property.

rhaas80 commented on config/chime_upgrade_network.j2:
```diff
@@ -0,0 +1,208 @@
+    kotekan_stage: cudaProcess
+    gpu_id: 0
+    buffer_depth: gpu_buffer_depth
+    frame_arrival_period: 390625/samples_per_frame
```
comment on magic number "390625"?

andrerenard replied:
Added a comment, this could maybe be dervied from the Telescope object in the code instead of the config, but there are cases it's nice to be able to set this independent of the F-engine parameters (e.g. in testing).

rhaas80 commented on config/chime_upgrade_network.j2:
```diff
@@ -0,0 +1,208 @@
+      gpu_mem_output: blocked_voltage_buffer
+
+
+# Hex dump the host_voltage_buffer_12 (for debugging)
```
12 is just any random buffer or special?

andrerenard replied:
Just a random buffer so I could peek at the data.  I'll remove this entirely, since it's not likely to be used by anyone else.

rhaas80 commented on lib/core/buffer.cpp:
```diff
@@ -595,7 +601,7 @@ uint8_t* Buffer::wait_for_empty_frame(const std::string& producer_name, const in
     // and forces a wait until that buffer has been marked as empty.
     DEBUG2("wait_for_empty_frame({:s}[{:d}]): waiting...", producer_name, ID);
-    print_full_status();
+    // print_full_status();
```
remove

andrerenard replied:
Actually I ment to leave that in.   Erik had added a bunch of status messages, but they overwhelm the output for these large pipelines.   Hence my other changes to allow one to selectively tune the logging on a per-buffer basis.

rhaas80 replied:
it's commented out though, you are saying to leave in the commented out code? Sometimes done as a hint for testers I guess.

andrerenard replied:
It's been changed back to uncommented, but I think because there is now no longer a change here, github is displaying in strangely.

rhaas80 commented on lib/core/ringbuffer.cpp:
```diff
@@ -233,7 +233,7 @@ void RingBuffer::finish_write(const std::string& name, const int inst, const std
            "produced bytes: {}",
            name, inst, group_digits(sz));
-    // print_full_status();
+    print_full_status();
```
is this `print_full_status()` correct here? produces output to screen one assumes.

andrerenard replied:
Yeah it produces data to screen.  Was commented out I assume because it was producing too much output, should be ok to have it back now with the logging changes.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+
+    signal_buffer = dynamic_cast<RingBuffer*>(
+        host_buffers.get_generic_buffer(config.get<std::string>(unique_name, "signal_buf")));
+    assert(signal_buffer);
```
add an error message, at least something like:

```
assert(signal_buffer && "Buffer given by signal_buf is not a RingBuffer");
```

andrerenard replied:
Done.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+}
+
+cudaCopyNtoRingbuffer::~cudaCopyNtoRingbuffer() {
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
```
why use `for(int i` type looping here by range-based loops before? The index `i` is not used for anything.

andrerenard replied:
Likely I was using i at some point, and never updated it.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+cudaCopyNtoRingbuffer::~cudaCopyNtoRingbuffer() {
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
+        Buffer* buf = in_buffers[i];
+        if (buf && buf->frame_size) {
```
for `buf` to be `NULL` line 34 above must not have triggered, so this would be a bug, yes? so instead use an `assert(buf)` maybe?

andrerenard replied:
No harm in that check.

andrerenard replied:
Also yeah only need to check frame_size here + an assert.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+
+int cudaCopyNtoRingbuffer::wait_on_precondition() {
+    // Wait for all input buffers to have data
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
```
same comment about `int i` vs range based loop as above

andrerenard replied:
Done.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+int cudaCopyNtoRingbuffer::wait_on_precondition() {
+    // Wait for all input buffers to have data
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
+        Buffer* buf = in_buffers[i];
```
I note no check of `buff == NULL` here.

andrerenard replied:
Done.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+        DEBUG("Waiting for input data frame {} from buffer {}", gpu_frame_id, buf->buffer_name);
+        uint8_t* frame = buf->wait_for_full_frame(unique_name, gpu_frame_id % buf->num_frames);
+        if (!frame)
+            return -1;
```
really a `return` or a `continue` instead (there could be valid frames after this one, could there?)

andrerenard replied:
I have not coded any logic to track frames in the sequency which are missing.  So yes, I want to return here and not attempt to transfer a partially filled frame to the GPU.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+        signal_buffer->wait_for_writable(unique_name, instance_num, total_input_size);
+    DEBUG("Ringbuffer space for writing is now available");
+    if (!val.has_value())
+        return -1;
```
explicitly invalidate `output_cursor`?

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+
+cudaEvent_t cudaCopyNtoRingbuffer::execute(cudaPipelineState& pipestate,
+                                           const std::vector<cudaEvent_t>& pre_events) {
+    (void)pipestate;
```
instead of C style cast to void, use

```
udaEvent_t cudaCopyNtoRingbuffer::execute(cudaPipelineState& /*pipestate*/,
                                           const std::vector<cudaEvent_t>& pre_events) {
```

instead?

andrerenard replied:
Done.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+    // Merge all input buffers into the ringbuffer sequentially
+    size_t offset = 0;
+    std::vector<std::shared_ptr<chordMetadata>> metas;
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
```
comment about looping

andrerenard replied:
Done.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+    size_t offset = 0;
+    std::vector<std::shared_ptr<chordMetadata>> metas;
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
+        Buffer* buf = in_buffers[i];
```
no check for `NULL`

andrerenard replied:
Done.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
+        Buffer* buf = in_buffers[i];
+        int buf_index = gpu_frame_id % buf->num_frames;
+        void* host_memory_frame = (void*)buf->frames[buf_index];
```
is this cast required? Doesn't even C++ let you cast any pointer to `void*` automatically?

andrerenard replied:
I think the "modern/correct way" is to use `static_cast<void*>`, so switched to that. Although yes, the cast isn't strictly required.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+
+        // Merge metadata from all input buffers
+        // NB This is highly specific to CHIME.
+        for (uint i = 0; i < in_buffers.size(); ++i) {
```
comment about looping

rhaas80 replied:
`uint`?

andrerenard replied:
Yeah should be `size_t`, `i` is actually useful here.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+            if (!meta_in)
+                throw std::runtime_error(
+                    "cudaCopyNtoRingbuffer: input buffer has no chordMetadata");
+            if (i == 0) {
```
add comment about this being the first loop iteration?

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+                                        meta_ring->coarse_freq + in_buffers.size()),
+                       ", "));
+        first_time = false;
+    }
```
`metas` is not used.

andrerenard replied:
Ah yes, removed.

rhaas80 commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+void cudaCopyNtoRingbuffer::finalize_frame() {
+    DEBUG("finalize_frame() for frame {}, releasing metadata on GPU output buffer {}", gpu_frame_id,
+          _gpu_mem_output);
+    for (size_t i = 0; i < in_buffers.size(); ++i) {
```
looping

andrerenard replied:
Done.

rhaas80 commented on lib/cuda/cudaCopyToRingbuffer.cpp:
```diff
@@ -49,6 +49,8 @@ cudaCopyToRingbuffer::cudaCopyToRingbuffer(Config& config, const std::string& un
             uint flags;
             // only register the memory if it isn't already...
+            // TODO This logic doesn't work for host buffers with more then num_gpu_frames frames
+            // (which is likely)
```
should I be worried?

andrerenard replied:
Actually I think all this register stuff can be removed.  The framework level calls to `cudaProcess::register_host_memory()` in the buffer mapping should take care of it...  If they aren't then as currently written this would cause a performance issue for the frames that aren't captured by this logic when the CPU and GPU frames have different lenghts.   I'm just going to remove it entirely, if it causes performance issues some place, then that's a bug and we should address it at the framework level, and not make the copy process deal with it.

rhaas80 commented on lib/dpdk/dpdkCore.cpp:
```diff
@@ -73,29 +70,17 @@ dpdkCore::dpdkCore(Config& config, const string& unique_name, bufferContainer& b
     memset((void*)&port_conf, 0, sizeof(struct rte_eth_conf));
-    uint32_t max_rx_pkt_len = config.get_default<uint32_t>(unique_name, "max_rx_pkt_len", 5000);
-#ifndef OLD_DPDK
+    uint32_t max_rx_pkt_len = config.get_default<uint32_t>(unique_name, "max_rx_pkt_len", 9000);
```
has this change in default been reflected in the doxygen docs?

rhaas80 commented on lib/dpdk/dpdkCore.hpp:
```diff
@@ -140,6 +140,7 @@ class dpdkRXhandler : public kotekan::kotekanLogging {
  * @conf   jumbo_frame      Bool. Default true Enable support for Jumbo frames
```
incorrect default for `max_rx_pkt_len` documented (see comment above)

rhaas80 replied:
`max_rx_pkt_len Int. Default 5000` still incorrect.

rhaas80 commented on lib/dpdk/iceBoardShuffle.hpp:
```diff
@@ -43,6 +44,8 @@
  *
  * @conf  fpga_dataset          String. The dataset ID for the data being received from
  *                              the F-engine.
+ * @conf  link_group            Int.  The link group this handler is a member of.
```
unsigned int I assume actually?

andrerenard replied:
Yes.

rhaas80 commented on lib/dpdk/iceBoardShuffle.hpp:
```diff
@@ -143,10 +146,19 @@ class iceBoardShuffle : public iceBoardHandler {
-    /// shuffle operations. In that case this will need to be factored out.
-    static ice_stream_id_t all_stream_ids[shuffle_size];
+    /// There are 4 stream_ids per link group.
+    inline static ice_stream_id_t all_stream_ids[link_group_max][shuffle_size];
```
I am confused. What does `inline` on a data member do?

andrerenard replied:
That's done to avoid needing an out of class definition (which was removed in this PR).   There are likely more elegant ways to do this, but since the DPDK code needs to be extremely efficient, I wanted to avoid requiring dynamic structures...

rhaas80 commented on lib/dpdk/iceBoardShuffle.hpp:
```diff
@@ -354,24 +379,33 @@ inline bool iceBoardShuffle::check_stream_id() {
+                      all_stream_ids[link_group][j].slot_id, all_stream_ids[link_group][j].link_id,
+                      port);
+            }
             FATAL_ERROR("One of the link_ids or slot_ids don't match! There is a cabling problem.");
```
At least one of...

andrerenard replied:
Done.

rhaas80 commented on lib/dpdk/iceBoardShuffle.hpp:
```diff
@@ -407,22 +441,58 @@ inline bool iceBoardShuffle::advance_frames(uint64_t new_seq, bool first_time) {
             tmp_stream_id.unused = i;
-            ice_set_stream_id_t(out_bufs[i], out_buf_frame_ids[i], tmp_stream_id);
+            tmp_stream_id.crate_id = tmp_stream_id.crate_id % 2;
+            meta->coarse_freq[0] = tel.to_freq_id(ice_encode_stream_id(tmp_stream_id));
```
single frequency only?

andrerenard replied:
Yep, each output frame is only one frequency.  This isn't something that's likely to change,    `iceBoardShuffle` is very specific to the CHIME F-engine design.

rhaas80 commented on lib/dpdk/iceBoardShuffle.hpp:
```diff
@@ -433,18 +503,43 @@ inline bool iceBoardShuffle::advance_frames(uint64_t new_seq, bool first_time) {
+        meta->type = kotekan::uint8;
+
+        std::strncpy(meta->dim_name[0], "T", sizeof meta->dim_name[0]);
+        meta->dim[0] = lost_samples_buf->frame_size; // One byte per time sample
```
no name for the data?

andrerenard replied:
Done.

rhaas80 commented on lib/dpdk/iceBoardShuffle.hpp:
```diff
@@ -433,18 +503,43 @@ inline bool iceBoardShuffle::advance_frames(uint64_t new_seq, bool first_time) {
+        meta->type = kotekan::uint8;
+
+        std::strncpy(meta->dim_name[0], "T", sizeof meta->dim_name[0]);
+        meta->dim[0] = lost_samples_buf->frame_size; // One byte per time sample
```
not setting `dims`, maybe other fields missing.

andrerenard replied:
Thanks for catching that.

andrerenard commented on lib/cuda/cudaCopyNtoRingbuffer.cpp:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    // We only need to set the metadata on the ring buffer once.
+    if (first_time && instance_num == 0) {
```
Thinking about this more we should implement a cross-check of the metadata on every call to this function.

andrerenard committed:
Address review comments

andrerenard commented:
Comments mostly addressed, need to run some tests to make sure I didn't break the system performance by removing the register logic.

andrerenard committed:
clang-format

andrerenard committed:
Add metadata consistency checking logic

andrerenard commented:
I think I've addressed all the comments.  I also added some metadata consistency checking logic, and confirmed that my removal of the cuda memory registration logic was safely handled already in the framework buffer mapping logic.  Ready for re-review.

rhaas80 commented:
https://github.com/kotekan/kotekan/pull/1302/files#r2403361840 remains unresolved. Fine otherwise.

andrerenard committed:
Fix max packet size in docs and config
